[PyTorch Edge] Backport function for defaults args with out args, flag on

cccclai opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #63651

1. Enable support for operators with default args and out args. For `torch.add(x, h, out=x)`, the number of specified arguments will be 3 instead of 4.
2. Bump bytecode version from 6 to 7
3. Implement backport_v7_to_v6 function. Also slightly refactor the local_thread to allow re-emit operators.
4. unittest to cover backport function
5. Update expect result from 4 to 3 in unit test DefaultArgsWithOutArg to cover the number of specified arguments.



Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080/)

Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080)

Commit:
[PyTorch Edge] Backport function for defaults args with out args



Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080/)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Backport function for defaults args with out args, flag on"


1. Enable support for operators with default args and out args. For `torch.add(x, h, out=x)`, the number of specified arguments will be 3 instead of 4.
2. Bump bytecode version from 6 to 7
3. Implement backport_v7_to_v6 function. Also slightly refactor the local_thread to allow re-emit operators.
4. unittest to cover backport function
5. Update expect result from 4 to 3 in unit test DefaultArgsWithOutArg to cover the number of specified arguments.



Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080/)

Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Backport function for defaults args with out args, flag on"


1. Enable support for operators with default args and out args. For `torch.add(x, h, out=x)`, the number of specified arguments will be 3 instead of 4.
2. Bump bytecode version from 6 to 7
3. Implement backport_v7_to_v6 function. Also slightly refactor the local_thread to allow re-emit operators.
4. unittest to cover backport function
5. Update expect result from 4 to 3 in unit test DefaultArgsWithOutArg to cover the number of specified arguments.



Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080/)

Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080)

[ghstack-poisoned]

tugsbayasgalan commented on torch/csrc/jit/mobile/backport_manager.cpp:
```diff
@@ -349,6 +352,95 @@ std::stringstream backport_v6_to_v5(std::stringstream& input_model_stream) {
+std::stringstream backport_v7_to_v6(std::stringstream& input_model_stream) {
+  std::shared_ptr<IStreamAdapter> rai =
+      std::make_shared<IStreamAdapter>(&input_model_stream);
+  auto reader = std::make_shared<PyTorchStreamReader>(rai);
```
Just nit: The initial part of this method seems similar to backport_v6_to_v5. Should we make it into an utility function?

cccclai replied:
Good point! Resolve in the latest commit. Create a function to re-emit the model, and the input is model_stream + different mode.

Commit:
Update on "[PyTorch Edge] Backport function for defaults args with out args, flag on"


1. Enable support for operators with default args and out args. For `torch.add(x, h, out=x)`, the number of specified arguments will be 3 instead of 4.
2. Bump bytecode version from 6 to 7
3. Implement backport_v7_to_v6 function. Also slightly refactor the local_thread to allow re-emit operators.
4. unittest to cover backport function
5. Update expect result from 4 to 3 in unit test DefaultArgsWithOutArg to cover the number of specified arguments.



Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080/)

Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Backport function for defaults args with out args, flag on"


1. Enable support for operators with default args and out args. For `torch.add(x, h, out=x)`, the number of specified arguments will be 3 instead of 4.
2. Bump bytecode version from 6 to 7
3. Implement backport_v7_to_v6 function. Also slightly refactor the local_thread to allow re-emit operators.
4. unittest to cover backport function
5. Update expect result from 4 to 3 in unit test DefaultArgsWithOutArg to cover the number of specified arguments.



Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080/)

Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080)

[ghstack-poisoned]

raziel commented on test/cpp/jit/test_lite_interpreter.cpp:
```diff
@@ -724,7 +725,10 @@ TEST(LiteInterpreterTest, BackPortByteCodeModelAllVersions) {
-      return (x1, x2, x3)
+      x = 2 * torch.ones(1)
+      h = torch.ones(1)
+      torch.add(x, h, out=x)
```
since you're expanding the model used to test all versions, maybe add a comment to this line since it's key to test v7->v6 backport

raziel commented on torch/csrc/jit/mobile/backport_manager.cpp:
```diff
@@ -96,115 +98,7 @@ void get_model_stream(PyTorchStreamReader& reader, std::stringstream& out) {
-}
-
-void writeArchiveV5(
+void write_archive_v5_to_v7(
```
v5 to v7? What of this function is version specific?

raziel commented on torch/csrc/jit/mobile/backport_manager.cpp:
```diff
@@ -11,6 +11,7 @@
 #include <torch/csrc/jit/serialization/pickler.h>
 #include <cstddef>
 #include <sstream>
+#include "ATen/core/interned_strings.h"
```
<ATen/core/interned_strings.h> and move up?

raziel commented on torch/csrc/jit/mobile/backport_manager.cpp:
```diff
@@ -327,28 +227,162 @@ std::stringstream backport_v6_to_v5(std::stringstream& input_model_stream) {
+} // namespace
+
+/*
+ To add next backport function, for example, backport_vn_to_vn-1, create an
```
I'm confused by the code organization: Why is this comment in the middle? There's also a bunch of opening and closing of namespace.

Can we instead use this structure:

1. The kBytecodeVersionVX constants
2. All the utility functions.
3. All the backport_vN_to_vN-1 functions.
4. BytecodeBackportFunction and BackportManager.

Right now it's hard to easily understand what are utility functions and what are the backport_vN_to_vN-1 entry points.

Also, rather than calling methods like write_archive_v5_to_v7, can we find better names and/or just document them?

Also, it seems wrong that you have what seem like general utilities (e.g. re_emit_model(...) ) which even take a version number but really don't work for all versions. In that case it would be best to break down those method into general utilities that the backport_vN_to_vN-1 functions call directly.
It may seem like we will have some repeated code but the general code will be so much more maintainable since you won't have thing like re_emit_model() that seems generic but it's not, nor functions like write_archive_v5_to_v7 that now are specialized to certain versions.

In summary:

* better error in smaller but generic utility functions at the cost of some duplicated code in backport_vN_to_vN-1.
* ideally, no version specific names/logic in the utilities --i.e. make them about their functionality, rather than tying them to the concept of a particular version.

That will also male backport_vN_to_vN-1 self documenting since we will be able to read at a high level what the backporting is doing, rather than traverse the current spaghetti code.

raziel commented on torch/csrc/jit/mobile/backport_manager.cpp:
```diff
@@ -327,28 +227,162 @@ std::stringstream backport_v6_to_v5(std::stringstream& input_model_stream) {
+  writer.writeRecord(fname, data.data(), data.size());
+}
+
+std::stringstream backport_v5_to_v4(std::stringstream& input_model_stream) {
```
I should have being stricter about this, but please let's document each of this backport functions: what was the version bump about, and what the backport is accomplishing. It's less important about the code details, since per my comment above my suggestions would make the code more self documenting.

raziel commented on torch/csrc/jit/mobile/backport_manager.cpp:
```diff
@@ -327,28 +227,162 @@ std::stringstream backport_v6_to_v5(std::stringstream& input_model_stream) {
   return ouput_model_stream;
 }
+
+std::stringstream backport_v7_to_v6(std::stringstream& input_model_stream) {
```
Eg. documentation:

Backport function bytecode v7 that introduced support for operators with out arguments.
Previously, in v6, operators with out arguments forced the serialization of all arguments in the schema, even when optional arguments were not provided (as they had default values).
Currently, in v7, operators are aware of out arguments being present in the schema (always appended), allowing the serialization of only required arguments (as default values will be provided by the runtime).
The bump was needed because the v7 bytecode specifies less arguments for ops with out arguments in the schema, since the runtime code is now able to query whether an argument is of type "out" and insert the necessary default values in the right order in the interpreter stack (i.e. before the out arguments).
For example schema is: torch.add(x, h, alpha=1.0, out=x)
So, if the program defines: torch.add(x, h, out=x)
Previously, in v6, we serialized the bytecode to contain all 4 arguments.
Currently, in v7, we serialize the bytecode with only 3 arguments, since alpha is optional and has a default value that the runtime will push in the stack.
Thus, the backport is necessary such that the bytecode contains all the arguments as before.

cccclai commented on caffe2/serialize/versions.h:
```diff
@@ -85,7 +85,7 @@ static_assert(kProducedBytecodeVersion >= kProducedFileFormatVersion,
 // handle an updated operator.
 constexpr uint64_t kMinSupportedBytecodeVersion = 0x3L;
-constexpr uint64_t kMaxSupportedBytecodeVersion = 0x6L;
+constexpr uint64_t kMaxSupportedBytecodeVersion = 0x7L;
```
That's true. I should bump the `kMaxSupportedBytecodeVersion` in previous pr.

Commit:
Update on "[PyTorch Edge] Backport function for defaults args with out args, flag on"


1. Enable support for operators with default args and out args. For `torch.add(x, h, out=x)`, the number of specified arguments will be 3 instead of 4.
2. Bump bytecode version from 6 to 7
3. Implement backport_v7_to_v6 function. Also slightly refactor the local_thread to allow re-emit operators.
4. unittest to cover backport function
5. Update expect result from 4 to 3 in unit test DefaultArgsWithOutArg to cover the number of specified arguments.



Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080/)

Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Backport function for defaults args with out args, flag on"


1. Enable support for operators with default args and out args. For `torch.add(x, h, out=x)`, the number of specified arguments will be 3 instead of 4.
2. Bump bytecode version from 6 to 7
3. Implement backport_v7_to_v6 function. Also slightly refactor the local_thread to allow re-emit operators.
4. unittest to cover backport function
5. Update expect result from 4 to 3 in unit test DefaultArgsWithOutArg to cover the number of specified arguments.



Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080/)

Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Backport function for defaults args with out args, flag on"


1. Enable support for operators with default args and out args. For `torch.add(x, h, out=x)`, the number of specified arguments will be 3 instead of 4.
2. Bump bytecode version from 6 to 7
3. Implement backport_v7_to_v6 function. Also slightly refactor the local_thread to allow re-emit operators.
4. unittest to cover backport function
5. Update expect result from 4 to 3 in unit test DefaultArgsWithOutArg to cover the number of specified arguments.



Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080/)

Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Backport function for defaults args with out args, flag on"


1. Enable support for operators with default args and out args. For `torch.add(x, h, out=x)`, the number of specified arguments will be 3 instead of 4.
2. Bump bytecode version from 6 to 7
3. Implement backport_v7_to_v6 function. Also slightly refactor the local_thread to allow re-emit operators.
4. unittest to cover backport function
5. Update expect result from 4 to 3 in unit test DefaultArgsWithOutArg to cover the number of specified arguments.



Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080/)

Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Backport function for defaults args with out args, flag on"


1. Enable support for operators with default args and out args. For `torch.add(x, h, out=x)`, the number of specified arguments will be 3 instead of 4.
2. Bump bytecode version from 6 to 7
3. Implement backport_v7_to_v6 function. Also slightly refactor the local_thread to allow re-emit operators.
4. unittest to cover backport function
5. Update expect result from 4 to 3 in unit test DefaultArgsWithOutArg to cover the number of specified arguments.



Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080/)

Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Backport function for defaults args with out args, flag on"


1. Enable support for operators with default args and out args. For `torch.add(x, h, out=x)`, the number of specified arguments will be 3 instead of 4.
2. Bump bytecode version from 6 to 7
3. Implement backport_v7_to_v6 function. Also slightly refactor the local_thread to allow re-emit operators.
4. unittest to cover backport function
5. Update expect result from 4 to 3 in unit test DefaultArgsWithOutArg to cover the number of specified arguments.



Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080/)

Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Backport function for defaults args with out args, flag on"


1. Enable support for operators with default args and out args. For `torch.add(x, h, out=x)`, the number of specified arguments will be 3 instead of 4.
2. Bump bytecode version from 6 to 7
3. Implement backport_v7_to_v6 function. Also slightly refactor the local_thread to allow re-emit operators.
4. unittest to cover backport function
5. Update expect result from 4 to 3 in unit test DefaultArgsWithOutArg to cover the number of specified arguments.



Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080/)

Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[PyTorch Edge] Backport function for defaults args with out args, flag on"


1. Enable support for operators with default args and out args. For `torch.add(x, h, out=x)`, the number of specified arguments will be 3 instead of 4.
2. Bump bytecode version from 6 to 7
3. Implement backport_v7_to_v6 function. Also slightly refactor the local_thread to allow re-emit operators.
4. unittest to cover backport function
5. Update expect result from 4 to 3 in unit test DefaultArgsWithOutArg to cover the number of specified arguments.



Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080/)

Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Backport function for defaults args with out args, flag on"


1. Enable support for operators with default args and out args. For `torch.add(x, h, out=x)`, the number of specified arguments will be 3 instead of 4.
2. Bump bytecode version from 6 to 7
3. Implement backport_v7_to_v6 function. Also slightly refactor the local_thread to allow re-emit operators.
4. unittest to cover backport function
5. Update expect result from 4 to 3 in unit test DefaultArgsWithOutArg to cover the number of specified arguments.



Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080/)

Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Backport function for defaults args with out args, flag on"


1. Enable support for operators with default args and out args. For `torch.add(x, h, out=x)`, the number of specified arguments will be 3 instead of 4.
2. Bump bytecode version from 6 to 7
3. Implement backport_v7_to_v6 function. Also slightly refactor the local_thread to allow re-emit operators.
4. unittest to cover backport function
5. Update expect result from 4 to 3 in unit test DefaultArgsWithOutArg to cover the number of specified arguments.



Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080/)

Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080)

[ghstack-poisoned]

tugsbayasgalan reviewed (approved):
From TS side, lgtm!

iseeyuan reviewed (commented):
LGTM. Please make sure to test all production models:
1. All re-exported to v7. Load & run.
2. All v7 models back ported to v6. Load & run.
The process is not purely automatic now, but it's worth to make sure we don't introduce bugs, especially generate wrong models for this PR. Since there were scripts when we tested v5 & v6, it may save some time. 

The PR below (#63540) can be landed before this one.

I'll leave @raziel to call since he made a lot of comments on previous version.

Commit:
Update on "[PyTorch Edge] Backport function for defaults args with out args, flag on"


1. Enable support for operators with default args and out args. For `torch.add(x, h, out=x)`, the number of specified arguments will be 3 instead of 4.
2. Bump bytecode version from 6 to 7
3. Implement backport_v7_to_v6 function. Also slightly refactor the local_thread to allow re-emit operators.
4. unittest to cover backport function
5. Update expect result from 4 to 3 in unit test DefaultArgsWithOutArg to cover the number of specified arguments.



Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080/)

Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080)

[ghstack-poisoned]

cccclai commented:
> LGTM. Please make sure to test all production models:
> 
> 1. All re-exported to v7. Load & run.
> 2. All v7 models back ported to v6. Load & run.
>    The process is not purely automatic now, but it's worth to make sure we don't introduce bugs, especially generate wrong models for this PR. Since there were scripts when we tested v5 & v6, it may save some time.
> 
> The PR below (#63540) can be landed before this one.
> 
> I'll leave @raziel to call since he made a lot of comments on previous version.

Good point. Let me see if I can put up a diff to make the process more automate.

raziel reviewed (approved):
Thanks for cleaning it up.
Some small comments on the code structure.

cccclai commented on torch/csrc/jit/mobile/backport_manager.cpp:
```diff
@@ -127,7 +272,40 @@ void get_model_stream(PyTorchStreamReader& reader, std::stringstream& out) {
 */
 
-// The functions needed for backport model from v5 to v4.
+/*
```
Thanks for the detail notes and explanation. Address the comments in the latest commit.

Commit:
Update on "[PyTorch Edge] Backport function for defaults args with out args, flag on"


1. Enable support for operators with default args and out args. For `torch.add(x, h, out=x)`, the number of specified arguments will be 3 instead of 4.
2. Bump bytecode version from 6 to 7
3. Implement backport_v7_to_v6 function. Also slightly refactor the local_thread to allow re-emit operators.
4. unittest to cover backport function
5. Update expect result from 4 to 3 in unit test DefaultArgsWithOutArg to cover the number of specified arguments.



Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080/)

Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080)

[ghstack-poisoned]

cccclai commented:
Thanks everyone for the detailed review. Test prod models and they are all good.

Commit:
Update on "[PyTorch Edge] Backport function for defaults args with out args, flag on"


1. Enable support for operators with default args and out args. For `torch.add(x, h, out=x)`, the number of specified arguments will be 3 instead of 4.
2. Bump bytecode version from 6 to 7
3. Implement backport_v7_to_v6 function. Also slightly refactor the local_thread to allow re-emit operators.
4. unittest to cover backport function
5. Update expect result from 4 to 3 in unit test DefaultArgsWithOutArg to cover the number of specified arguments.



Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080/)

Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Backport function for defaults args with out args, flag on"


1. Enable support for operators with default args and out args. For `torch.add(x, h, out=x)`, the number of specified arguments will be 3 instead of 4.
2. Bump bytecode version from 6 to 7
3. Implement backport_v7_to_v6 function. Also slightly refactor the local_thread to allow re-emit operators.
4. unittest to cover backport function
5. Update expect result from 4 to 3 in unit test DefaultArgsWithOutArg to cover the number of specified arguments.



Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080/)

Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Backport function for defaults args with out args, flag on"


1. Enable support for operators with default args and out args. For `torch.add(x, h, out=x)`, the number of specified arguments will be 3 instead of 4.
2. Bump bytecode version from 6 to 7
3. Implement backport_v7_to_v6 function. Also slightly refactor the local_thread to allow re-emit operators.
4. unittest to cover backport function
5. Update expect result from 4 to 3 in unit test DefaultArgsWithOutArg to cover the number of specified arguments.



Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080/)

Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Backport function for defaults args with out args, flag on"


1. Enable support for operators with default args and out args. For `torch.add(x, h, out=x)`, the number of specified arguments will be 3 instead of 4.
2. Bump bytecode version from 6 to 7
3. Implement backport_v7_to_v6 function. Also slightly refactor the local_thread to allow re-emit operators.
4. unittest to cover backport function
5. Update expect result from 4 to 3 in unit test DefaultArgsWithOutArg to cover the number of specified arguments.



Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080/)

Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Backport function for defaults args with out args, flag on"


1. Enable support for operators with default args and out args. For `torch.add(x, h, out=x)`, the number of specified arguments will be 3 instead of 4.
2. Bump bytecode version from 6 to 7
3. Implement backport_v7_to_v6 function. Also slightly refactor the local_thread to allow re-emit operators.
4. unittest to cover backport function
5. Update expect result from 4 to 3 in unit test DefaultArgsWithOutArg to cover the number of specified arguments.



Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080/)

Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Backport function for defaults args with out args, flag on"


1. Enable support for operators with default args and out args. For `torch.add(x, h, out=x)`, the number of specified arguments will be 3 instead of 4.
2. Bump bytecode version from 6 to 7
3. Implement backport_v7_to_v6 function. Also slightly refactor the local_thread to allow re-emit operators.
4. unittest to cover backport function
5. Update expect result from 4 to 3 in unit test DefaultArgsWithOutArg to cover the number of specified arguments.



Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080/)

Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Backport function for defaults args with out args, flag on"


1. Enable support for operators with default args and out args. For `torch.add(x, h, out=x)`, the number of specified arguments will be 3 instead of 4.
2. Bump bytecode version from 6 to 7
3. Implement backport_v7_to_v6 function. Also slightly refactor the local_thread to allow re-emit operators.
4. unittest to cover backport function
5. Update expect result from 4 to 3 in unit test DefaultArgsWithOutArg to cover the number of specified arguments.



Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080/)

Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Backport function for defaults args with out args, flag on"


1. Enable support for operators with default args and out args. For `torch.add(x, h, out=x)`, the number of specified arguments will be 3 instead of 4.
2. Bump bytecode version from 6 to 7
3. Implement backport_v7_to_v6 function. Also slightly refactor the local_thread to allow re-emit operators.
4. unittest to cover backport function
5. Update expect result from 4 to 3 in unit test DefaultArgsWithOutArg to cover the number of specified arguments.



Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080/)

Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Backport function for defaults args with out args, flag on"


1. Enable support for operators with default args and out args. For `torch.add(x, h, out=x)`, the number of specified arguments will be 3 instead of 4.
2. Bump bytecode version from 6 to 7
3. Implement backport_v7_to_v6 function. Also slightly refactor the local_thread to allow re-emit operators.
4. unittest to cover backport function
5. Update expect result from 4 to 3 in unit test DefaultArgsWithOutArg to cover the number of specified arguments.



Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080/)

Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Backport function for defaults args with out args, flag on"


1. Enable support for operators with default args and out args. For `torch.add(x, h, out=x)`, the number of specified arguments will be 3 instead of 4.
2. Bump bytecode version from 6 to 7
3. Implement backport_v7_to_v6 function. Also slightly refactor the local_thread to allow re-emit operators.
4. unittest to cover backport function
5. Update expect result from 4 to 3 in unit test DefaultArgsWithOutArg to cover the number of specified arguments.



Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080/)

Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Backport function for defaults args with out args, flag on"


1. Enable support for operators with default args and out args. For `torch.add(x, h, out=x)`, the number of specified arguments will be 3 instead of 4.
2. Bump bytecode version from 6 to 7
3. Implement backport_v7_to_v6 function. Also slightly refactor the local_thread to allow re-emit operators.
4. unittest to cover backport function
5. Update expect result from 4 to 3 in unit test DefaultArgsWithOutArg to cover the number of specified arguments.



Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080/)

Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Backport function for defaults args with out args, flag on"


1. Enable support for operators with default args and out args. For `torch.add(x, h, out=x)`, the number of specified arguments will be 3 instead of 4.
2. Bump bytecode version from 6 to 7
3. Implement backport_v7_to_v6 function. Also slightly refactor the local_thread to allow re-emit operators.
4. unittest to cover backport function
5. Update expect result from 4 to 3 in unit test DefaultArgsWithOutArg to cover the number of specified arguments.



Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080/)

Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Backport function for defaults args with out args, flag on"


1. Enable support for operators with default args and out args. For `torch.add(x, h, out=x)`, the number of specified arguments will be 3 instead of 4.
2. Bump bytecode version from 6 to 7
3. Implement backport_v7_to_v6 function. Also slightly refactor the local_thread to allow re-emit operators.
4. unittest to cover backport function
5. Update expect result from 4 to 3 in unit test DefaultArgsWithOutArg to cover the number of specified arguments.



Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080/)

Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Backport function for defaults args with out args, flag on"


1. Enable support for operators with default args and out args. For `torch.add(x, h, out=x)`, the number of specified arguments will be 3 instead of 4.
2. Bump bytecode version from 6 to 7
3. Implement backport_v7_to_v6 function. Also slightly refactor the local_thread to allow re-emit operators.
4. unittest to cover backport function
5. Update expect result from 4 to 3 in unit test DefaultArgsWithOutArg to cover the number of specified arguments.



Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080/)

Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Backport function for defaults args with out args, flag on"


1. Enable support for operators with default args and out args. For `torch.add(x, h, out=x)`, the number of specified arguments will be 3 instead of 4.
2. Bump bytecode version from 6 to 7
3. Implement backport_v7_to_v6 function. Also slightly refactor the local_thread to allow re-emit operators.
4. unittest to cover backport function
5. Update expect result from 4 to 3 in unit test DefaultArgsWithOutArg to cover the number of specified arguments.



Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080/)

Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Backport function for defaults args with out args, flag on"


1. Enable support for operators with default args and out args. For `torch.add(x, h, out=x)`, the number of specified arguments will be 3 instead of 4.
2. Bump bytecode version from 6 to 7
3. Implement backport_v7_to_v6 function. Also slightly refactor the local_thread to allow re-emit operators.
4. unittest to cover backport function
5. Update expect result from 4 to 3 in unit test DefaultArgsWithOutArg to cover the number of specified arguments.



Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080/)

Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Backport function for defaults args with out args, flag on"


1. Enable support for operators with default args and out args. For `torch.add(x, h, out=x)`, the number of specified arguments will be 3 instead of 4.
2. Bump bytecode version from 6 to 7
3. Implement backport_v7_to_v6 function. Also slightly refactor the local_thread to allow re-emit operators.
4. unittest to cover backport function
5. Update expect result from 4 to 3 in unit test DefaultArgsWithOutArg to cover the number of specified arguments.



Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080/)

Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Backport function for defaults args with out args, flag on"


1. Enable support for operators with default args and out args. For `torch.add(x, h, out=x)`, the number of specified arguments will be 3 instead of 4.
2. Bump bytecode version from 6 to 7
3. Implement backport_v7_to_v6 function. Also slightly refactor the local_thread to allow re-emit operators.
4. unittest to cover backport function
5. Update expect result from 4 to 3 in unit test DefaultArgsWithOutArg to cover the number of specified arguments.



Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080/)

Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Backport function for defaults args with out args, flag on"


1. Enable support for operators with default args and out args. For `torch.add(x, h, out=x)`, the number of specified arguments will be 3 instead of 4.
2. Bump bytecode version from 6 to 7
3. Implement backport_v7_to_v6 function. Also slightly refactor the local_thread to allow re-emit operators.
4. unittest to cover backport function
5. Update expect result from 4 to 3 in unit test DefaultArgsWithOutArg to cover the number of specified arguments.



Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080/)

Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Backport function for defaults args with out args, flag on"


1. Enable support for operators with default args and out args. For `torch.add(x, h, out=x)`, the number of specified arguments will be 3 instead of 4.
2. Bump bytecode version from 6 to 7
3. Implement backport_v7_to_v6 function. Also slightly refactor the local_thread to allow re-emit operators.
4. unittest to cover backport function
5. Update expect result from 4 to 3 in unit test DefaultArgsWithOutArg to cover the number of specified arguments.



Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080/)

Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Backport function for defaults args with out args, flag on"


1. Enable support for operators with default args and out args. For `torch.add(x, h, out=x)`, the number of specified arguments will be 3 instead of 4.
2. Bump bytecode version from 6 to 7
3. Implement backport_v7_to_v6 function. Also slightly refactor the local_thread to allow re-emit operators.
4. unittest to cover backport function
5. Update expect result from 4 to 3 in unit test DefaultArgsWithOutArg to cover the number of specified arguments.



Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080/)

Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Backport function for defaults args with out args, flag on"


1. Enable support for operators with default args and out args. For `torch.add(x, h, out=x)`, the number of specified arguments will be 3 instead of 4.
2. Bump bytecode version from 6 to 7
3. Implement backport_v7_to_v6 function. Also slightly refactor the local_thread to allow re-emit operators.
4. unittest to cover backport function
5. Update expect result from 4 to 3 in unit test DefaultArgsWithOutArg to cover the number of specified arguments.



Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080/)

Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Backport function for defaults args with out args, flag on"


1. Enable support for operators with default args and out args. For `torch.add(x, h, out=x)`, the number of specified arguments will be 3 instead of 4.
2. Bump bytecode version from 6 to 7
3. Implement backport_v7_to_v6 function. Also slightly refactor the local_thread to allow re-emit operators.
4. unittest to cover backport function
5. Update expect result from 4 to 3 in unit test DefaultArgsWithOutArg to cover the number of specified arguments.



Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080/)

Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080)

[ghstack-poisoned]

Commit:
Update on "[PyTorch Edge] Backport function for defaults args with out args, flag on"


1. Enable support for operators with default args and out args. For `torch.add(x, h, out=x)`, the number of specified arguments will be 3 instead of 4.
2. Bump bytecode version from 6 to 7
3. Implement backport_v7_to_v6 function. Also slightly refactor the local_thread to allow re-emit operators.
4. unittest to cover backport function
5. Update expect result from 4 to 3 in unit test DefaultArgsWithOutArg to cover the number of specified arguments.



Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080/)

Differential Revision: [D30454080](https://our.internmc.facebook.com/intern/diff/D30454080)

[ghstack-poisoned]
